discovery_engine_core/core/src/stack/ops/favourite.rs

```diff
@@ -0,0 +1,129 @@
+// Copyright 2022 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use std::{iter, sync::Arc};
+
+use async_trait::async_trait;
+use futures::stream::FuturesUnordered;
+use itertools::chain;
+use tokio::{sync::RwLock, task::JoinHandle};
+use uuid::Uuid;
+use xayn_ai::ranker::KeyPhrase;
+use xayn_discovery_engine_providers::{Article, Client, CommonQueryParts, HeadlinesQuery};
+
+use crate::{
+    document::{dedup_documents, Document, HistoricDocument},
+    engine::{EndpointConfig, GenericError},
+    stack::{
+        filters::{filter_semantically, ArticleFilter, CommonFilter, SemanticFilterConfig},
+        Id,
+    },
+};
+
+use super::{common::request_min_new_items, Ops};
+
+/// Stack operations customized for favourite news.
+pub(crate) struct FavouriteNews {
+    client: Arc<Client>,
+    sources: Arc<RwLock<Vec<String>>>,
```

Comment:
🎨 we could change this type to
```suggestion
    sources: Arc<RwLock<Arc<Vec<String>>>>,
```
(the endpoint config could already be changed to hold an `Arc<Vec<String>>`)

then in l.84 everytime we read the sources to get new items, the cloning will be cheap. i guess we should do the same for the excluded sources fields of the other stacks as well (maybe in another pr).

Reply:
I agree this would be better in a separate pr